src/metatrain/utils/custom_loss.py

```diff
@@ -0,0 +1,430 @@
+    """
+
+    initialized: bool = False
+
+    @abstractmethod
+    def initialize(
+        self, loss_fn: LossInterface, targets: Dict[str, TensorMap]
+    ) -> float:
+        """Compute and return the initial weight."""
+
+    @abstractmethod
+    def update(
+        self,
+        loss_fn: LossInterface,
+        predictions: Dict[str, TensorMap],
+        targets: Dict[str, TensorMap],
+    ) -> float:
+        """Update and return the new weight after a batch."""
+
+
+class EMAScheduler(WeightScheduler):
+    """
+    Exponential moving average scheduler for loss weights.
+    """
+
+    EPS = 1e-6
+    initialized: bool = False
+
+    def __init__(self, sliding_factor: Optional[float]) -> None:
+        self.sf: float = float(sliding_factor or 0.0)
+        self.weight: float = 1.0
+        self.initialized: bool = False
+
+    def initialize(
+        self, loss_fn: LossInterface, targets: Dict[str, TensorMap]
+    ) -> float:
+        if self.sf <= 0.0:
+            self.weight = 1.0
+        else:
+            name = loss_fn.target
+            grad = getattr(loss_fn, "gradient", None)
+            tm = targets[name]
+            if grad is None:
+                mean_tm = mts.mean_over_samples(tm, tm.sample_names)
+                baseline = TensorMap(
+                    keys=tm.keys,
+                    blocks=[
+                        mts.TensorBlock(
+                            samples=block.samples,
+                            components=block.components,
+                            properties=block.properties,
+                            values=torch.ones_like(block.values) * mean_block.values,
+                        )
+                        for block, mean_block in zip(tm, mean_tm)
+                    ],
+                )
+            else:
+                baseline = mts.zeros_like(tm)
+
+            val = loss_fn.compute({name: tm}, {name: baseline})
+            self.weight = float(val.clamp_min(self.EPS))
+        self.initialized = True
+        return self.weight
+
+    def update(
+        self,
+        loss_fn: LossInterface,
+        predictions: Dict[str, TensorMap],
+        targets: Dict[str, TensorMap],
+    ) -> float:
+        if self.sf <= 0.0:
+            return self.weight
+        err = loss_fn.compute(predictions, targets).detach().item()
+        new_weight = self.sf * self.weight + (1.0 - self.sf) * err
+        self.weight = max(new_weight, self.EPS)
+        return self.weight
+
+
+class ScheduledLoss(LossInterface):
+    """
+    Wrap a base :py:class:`LossInterface` with a :py:class:`WeightScheduler`.
+
+    After each compute, the scheduler updates the loss weight.
+    """
+
+    def __init__(
+        self,
+        base_loss: LossInterface,
+        scheduler: WeightScheduler,
+    ):
+        # Delegate attributes
+        self.base = base_loss
+        self.scheduler = scheduler
+        self.target = base_loss.target
+        self.reduction = base_loss.reduction
```

Comment:
If we're always composing / delegatin then it might be nicer to store the object and use it, e.g.
`self._base_loss = base_loss`

Reply:
Just in case we forget to delegate anything else later.